test/Prototypes/AU3.swift.gyb

```diff
@@ -814,6 +852,70 @@ extension String : Hashable {
   }
 }
 
+//===--- CString Interop --------------------------------------------------===//
+extension String {
+  /// Constructs a `String` having the same contents as `nulTerminatedUTF8`.
+  ///
+  /// - Parameter nulTerminatedUTF8: a sequence of contiguous UTF-8 encoded 
+  ///   bytes ending just before the first zero byte (NUL character).
+  init(cString nulTerminatedUTF8: UnsafePointer<CChar>) {
+    // FIXME: is UTF8 the right default way to interpret a C string?  Cocoa
+    // makes use of a "system encoding" that might be relevant.
+    
+    let count = UTF8._nullCodeUnitOffset(in: nulTerminatedUTF8)
+    content = nulTerminatedUTF8.withMemoryRebound(
+      to: UTF8.CodeUnit.self, capacity: count
+    ) {
+      Content(
+        utf16: _UnicodeViews(
+          UnsafeBufferPointer(start: $0, count: count),
+          UTF8.self).transcoded(to: UTF16.self))
+    }
+  }
+  
+  /// Constructs a `String` having the same contents as `nulTerminatedCodeUnits`.
+  ///
+  /// - Parameter nulTerminatedCodeUnits: a sequence of contiguous code units in
+  ///   the given `encoding`, ending just before the first zero code unit.
+  /// - Parameter encoding: describes the encoding in which the code units
+  ///   should be interpreted.
+  init<Encoding: UnicodeEncoding>(
+    cString nulTerminatedCodeUnits:
+    UnsafePointer<Encoding.EncodedScalar.Iterator.Element>,
+    encoding: Encoding.Type
+  )
+  where Encoding.EncodedScalar.Iterator.Element : UnsignedInteger {
+    let count = Encoding._nullCodeUnitOffset(in: nulTerminatedCodeUnits)
+    let source = UnsafeBufferPointer(
+      start: nulTerminatedCodeUnits, count: count)
+    content = Content(
+      utf16: _UnicodeViews(source, Encoding.self).transcoded(to: UTF16.self))
+  }
+ 
+  /// Invokes the given closure on the contents of the string, represented as a
+  /// pointer to a null-terminated sequence of UTF-8 code units.
+  func withCString<Result>(
+    _ body: (UnsafePointer<CChar>) throws -> Result) rethrows -> Result {
+    /*
+    // FIXME: add support for writing zero into the end of existing latin-1
+    // storage.
+    if let l = content._latin1, l.isASCII() {
+      return l.withExistingUnsafeBuffer(body)!
+    }
+    */
+    // FIXME: add a utf8 view to UnicodeContent so this can be faster in many
+    // cases.
+    var result = ContiguousArray<CChar>()
+    let source = Counted(content.utf8)
+    result.reserveCapacity(numericCast(source.count + 1))
+    result.append(contentsOf: source.lazy.map { numericCast($0) })
```

Comment:
Doesn't this need to be `Int8.init(bitPattern:)` rather than `numericCast`, which will trap for UTF8 values > 128?